### ISSUE TITLE
Numerical stability of float32 `torch.linalg.inv` and `torch.linalg.pinv` 

### ISSUE THREAD

**Issue by Redjest** [pytorch/pytorch#65557]
## 📚 Documentation

PyTorch's matrix inversion functions `torch.linalg.inv` and `torch.linalg.pinv` may be numerically unstable (known, documented). Working on explicit matrix models with PyTorch, I saw several times on different problems that float32 precision resulted in model divergence or poor performance while the simple casting to a double(float64) tensor before matrix inversion (either by `inv` or `pinv`) and back to 32 after fixed the issue.

I suggest doing one of the following:
- Add a note in the docs of the functions suggesting users facing this issue to try double precision
- Add internal casting to double functionality that may be turned on by an argument to the function (e.g. `double=True` that defaults to `False`)

Would be glad to implement either 😃    

cc @brianjo @mruberry @jianyuh @nikitaved @pearu @walterddr @IvanYashchuk @xwang233 @Lezcano

**Comment by mruberry** [comment]
Hey @Redjest, thanks for this suggestion.

I think we expect users to understand that using a more precise dtype may provide better results, so general warnings like "the accuracy of results will improve when using a more precise dtype" are not especially interesting. If, however, an operator is so numerically unstable that it frequently produces incorrect results in a dtype then we could consider upcasting the computation by default, but PyTorch is also very conservative about doing this because we like to allow users explicit control. Finally we could add a kwarg to optionally control the dtype the computation occurs in, and some other operators in PyTorch even have such dtypes, but the discoverability and utility of these kwargs is low so they're not that compelling today.

So while I agree with you completely and think it would be nice to have some notion of how numerically sensitive PyTorch operators are, updating the docs or adding a kwarg to inv and/or pinv doesn't seem like it'll do much. If we write a linear algebra note, however, that may be the best place to include the general guidance that linear algebra functions may be especially sensitive to precision. 

What do you think?

**Comment by Redjest** [comment]
Maybe describing the specific observations will help the discussion:
- Inverting an almost non-invertible (ill-conditioned) matrix, e.g. `linalg.inv(A)` for an A  with very small `det(A)`,  won't raise an exception as  `det(a)` is not strictly 0, but returns an invert matrix with exploding elements, very different from `pinv(A)`. This doesn't usually doesn't happen on double precision, either the matrix is detected as non invertible or the result invert is fine.
- Using many pseudoinverse matrices on the same model (e.g. implementing an RLS filter https://en.wikipedia.org/wiki/Recursive_least_squares_filter) results in useless models unless double precision is used.
- `pinv()` of some matrices of the form `D - I` when D is diagonal may result in some cases in a RuntimeError saying `svd_cpu: the updating process of SBDSDC did not converge` that occurs only on specific CPU architectures (e.g. in your hard to debug production env but not on your nice and cozy dev env). Also solved by double precision.

> ...upcasting the computation by default...

For all practical use cases I tried, eventually I needed double precision. But I agree that leaving it to the user to decide serves a greater audience. The question is if real world uses that can safely use f32 here actually exist, didn't try enough different problems to answer it.

> linear algebra functions may be especially sensitive to precision...

Not sure if all of them has this problem, personally encountered problems only with these two, but didn't try all of them.

**Comment by Lezcano** [comment]
Side note: As the documentation notes, you should almost never invert a matrix explicitly. You should prefer `torch.linalg.solve` because of these reasons. There are a number of cases when inverting and then multiplying is very badly conditioned, while solving it's fine (e.g. when the matrices A and B both have very small eigenvalues).

I agree with @mruberry in that many linalg functions are very sensitive to precision, with the inverse being one of the worst offenders, but closely followed  by `linalg.eig` / `linalg.eigvals`. Even more, when working with close-to-singular or badly conditioned matrices, any decomposition will struggle really.

**Comment by Redjest** [comment]
> You should prefer torch.linalg.solve

Of course 👍 

@Lezcano, so in your opinion should pytorch warn the users in the docs, do upcasting internally, add a kwarg for upcasting or is't just fine as is?

**Comment by Lezcano** [comment]
In my opinion it is fine as it is. Casting inside is not transparent and adding a kwarg is not necessary, as it's even shorter to perform the cast explicitly as: `torch.linalg.inv(X.double())`. As the zen goes:
>Explicit is better than implicit.
>There should be one-- and preferably only one --obvious way to do it.

About the docs, as mentioned above, this is more of a general linalg issue. You'd be surprised how difficult is to implement even [`torch.sum` ](https://www.youtube.com/watch?v=B1eFGn5nN84). Here we find a similar problem, only that since this is an inverse matrix function (again, like `eig` or other matrix decompositions), the errors are even larger. As mentioned by @mruberry , I think this would belong in a note on "general things to be aware of when dealing with linear algebra".

**Comment by Redjest** [comment]
Agree that `torch.linalg.inv(X.double())` is very elegant already. The reason I opened this issue is that sometimes, especially if your model is non-trivial or your code base is extensive, understanding that the bad results you get are due to a numerical precision problem may take a while. Hope to save others some debug time.

Thx for the link to prof. Higham's lecture (:

A note on numerical stability of linear algebra functions sounds good, it may also encourage people to contribute more useful linalg tips. From a user experience point of view I'd link to it from the doc pages of all affected functions as many users will probably look up the specific function and go from a google result page directly to the docs page of the function.

**Comment by Lezcano** [comment]
Closing this one as this is expected, not only in linalg but in many other applications of single-precision. We might follow-up in the future on this by doing a write-up of the usual caveats one may find when dealing with the linalg module. If you'd be interested in sometihng like this, please open a new issue indicating so to gauge general interest :)

**Comment by Redjest** [comment]
Sharing here the solution we used eventually, it may useful for others as the use case is common.

Given a pipeline on which data that goes thorough non-trivial pre-processing on CPUs and then used for training of  deep learning models:
- Do all the CPU side using NumPy. Less numerical issues, broad ecosystem that offers almost any algorithm you need, many data pipelining/parallelization frameworks (e.g. dask) have specific support for NumPy arrays.
- Convert to `torch.Tensor` right before sending data to GPUs

**Comment by Lezcano** [comment]
Just for the record, which functions were you missing, and which functions you found had better stability in NumPy than PyTorch?

**Comment by Redjest** [comment]
Numerical stability was better for linear algebra functions (probably due to the default of double precision, as both torch and numpy use LAPACK...). As for missing functions, the NumPy ecosystem has many domain specific functions that are probably just irrelevant to include in torch: DSP, statistical models, functions processing specific types of signals (e.g. ECG).

**Comment by IvanYashchuk** [comment]
> Numerical stability was better for linear algebra functions (probably due to the default of double-precision, as both torch and numpy use LAPACK...).

All NumPy linear algebra functions convert the data to double internally to do the computation and then cast the result back to the input type. In PyTorch, the current preference is to keep all dtype casting of inputs explicit and controlled by users.

**Comment by Redjest** [comment]
Single precision casts and half precision raises an exception
